src/t_zset.c

```diff
@@ -2184,6 +2188,165 @@ inline static void zunionInterAggregate(double *target, double val, int aggregat
     }
 }
 
+int dictGetMaxElementLength(dict *d) {
+    dictIterator *di;
+    dictEntry *de;
+    size_t maxelelen = 0;
+
+    di = dictGetIterator(d);
+
+    while((de = dictNext(di)) != NULL) {
+        sds ele = dictGetKey(de);
+        if (sdslen(ele) > maxelelen) maxelelen = sdslen(ele);
+    }
+
+    return maxelelen;
+}
+
+void zdiffAlgorithm1(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    int j;
+    zsetopval zval;
+    zskiplistNode *znode;
+    sds tmp;
+
+    /* With algorithm 1 it is better to order the sets to subtract
+     * by decreasing size, so that we are more likely to find
+     * duplicated elements ASAP. */
+    qsort(src+1,setnum-1,sizeof(zsetopsrc),zuiCompareByRevCardinality);
+
+    memset(&zval, 0, sizeof(zval));
+    zuiInitIterator(&src[0]);
+    while (zuiNext(&src[0],&zval)) {
+        double score, value;
+        int exists = 0;
+
+        score = zval.score;
+        if (isnan(score)) score = 0;
+
+        for (j = 1; j < setnum; j++) {
+            /* It is not safe to access the zset we are
+                * iterating, so explicitly check for equal object. */
+            if (src[j].subject == src[0].subject ||
+                    zuiFind(&src[j],&zval,&value)) {
+                exists = 1;
+                break;
+            }
+        }
+
+        if (!exists) {
+            tmp = zuiNewSdsFromValue(&zval);
+            znode = zslInsert(dstzset->zsl,score,tmp);
+            dictAdd(dstzset->dict,tmp,&znode->score);
+            if (sdslen(tmp) > *maxelelen) *maxelelen = sdslen(tmp);
+        }
+    }
+}
+
+
+void zdiffAlgorithm2(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    /* DIFF Algorithm 2:
+     *
+     * Add all the elements of the first set to the auxiliary set.
+     * Then remove all the elements of all the next sets from it.
+     *
+     * This is O(N) where N is the sum of all the elements in every
```

Comment:
do we need to mention the extra loop in dictGetMaxElementLength in this O formula?
if you're unsure too, i'll call in some Mathematician to the rescue. (this needs to also be mentioned in the docs).
and maybe updated in chooseDiffAlgorithm too

Reply:
It's been a long time since I formally analyzed any algorithm, so I'm not sure I correctly recall every detail about O notation and formulas, but I'm pretty sure it's not required, the N of the algorithm is always greater than the size of the resulting zset. But maybe it's worth mentioning it anyway? Not sure how strict we should be with the O notation here.

But this got me thinking about something else that I had forgotten: I copied this comment from the `t_set.c` file, but comparing other commands complexities, I believe this is wrong for t_zset.c, it seems the basic operations used by the algorithm have a non-constant cost (something like O(log N)). I'll check this again to see if this should be different.

Reply:
Just pushed some changes  to these comments with an updated algorithm complexity.